File: projekt/src/main/java/agh/ics/oop/model/Vector2d.java
package agh.ics.oop.model;

public record Vector2d(int x, int y) {

    @Override
    public String toString() {
        return "(%d,%d)".formatted(x, y);
    }

    public boolean precedes(Vector2d other) {
        return x <= other.x() && y <= other.y();
    }

    public boolean follows(Vector2d other) {
        return x >= other.x() && y >= other.y();
    }

    public Vector2d add(Vector2d other) {
        return new Vector2d(x + other.x(), y + other.y());
    }

    Vector2d upperRight(Vector2d other) {
        return new Vector2d(Math.max(x, other.x()), Math.max(y, other.y()));
    }

    Vector2d lowerLeft(Vector2d other) {
        return new Vector2d(Math.min(x, other.x()), Math.min(y, other.y()));
    }

    Vector2d subtract(Vector2d other) {
        return new Vector2d(x - other.x(), y - other.y());
    }

    Vector2d opposite() {
        return new Vector2d(-x, -y);
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null || obj.getClass() != this.getClass()) {
            return false;
        }
        var other = (Vector2d) obj;
        return this.x == other.x() && this.y == other.y();
    }

}

File: ProjektKotlina/src/main/kotlin/model/Vector2d.kt
package model

data class Vector2d(val x: Int, val y: Int) {

    override fun toString(): String = "($x,$y)"

    operator fun plus(other: Vector2d): Vector2d = Vector2d(x + other.x, y + other.y)

    operator fun minus(other: Vector2d): Vector2d = Vector2d(x - other.x, y - other.y)

    operator fun compareTo(other: Vector2d): Int {
        val precedes = (x <= other.x) && (y <= other.y)
        val follows = (x >= other.x) && (y >= other.y)
        return when {
            precedes && follows -> 0
            precedes -> -1
            else -> 1
        }
    }

    fun upperRight(other: Vector2d): Vector2d =
        Vector2d(maxOf(x, other.x), maxOf(y, other.y))

    fun lowerLeft(other: Vector2d): Vector2d =
        Vector2d(minOf(x, other.x), minOf(y, other.y))

    fun opposite(): Vector2d = Vector2d(-x, -y)

}

fun MapDirection.toUnitVector(): Vector2d = when (this) {
    MapDirection.NORTH -> Vector2d(0, 1)
    MapDirection.SOUTH -> Vector2d(0, -1)
    MapDirection.EAST -> Vector2d(1, 0)
    MapDirection.WEST -> Vector2d(-1, 0)
}